File: turtlesim_catch_them_all/src/turtle_controller.cpp
#include <cmath>
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "my_robot_interfaces/msg/turtle.hpp"
#include "my_robot_interfaces/msg/turtle_array.hpp"
#include "my_robot_interfaces/srv/catch_turtle.hpp"

using namespace std::placeholders;
using namespace std::chrono_literals ;

class TurtleControllerNode :public rclcpp::Node 
{
public :
    TurtleControllerNode() : Node("turtle_controller"), turtlesim_up_(false), catch_closest_turtle_first_(true)
    {
        catch_turtle_client_ = this->create_client<my_robot_interfaces::srv::CatchTurtle>("/catch_turtle") ;
        alive_turtles_subscrip_ = this->create_subscription<my_robot_interfaces::msg::TurtleArray>
            ("/alive_turtles", 10, std::bind(&TurtleControllerNode::callbackAliveTurtles, this, _1)) ;
        turtle_pose_ = this->create_subscription<turtlesim::msg::Pose> ("/turtle1/pose", 10, 
                       std::bind(&TurtleControllerNode::callbackTurtlepose, this, _1)) ;
        turtle_cmd_vel_ = this->create_publisher<geometry_msgs::msg::Twist> ("/turtle1/cmd_vel", 10) ;               
        control_loob_timer_ = this->create_wall_timer(0.001s, std::bind(&TurtleControllerNode::ControLoop, this)) ;  
    }
private :
 double getDistanceFromCurrentPose(my_robot_interfaces::msg::Turtle turtle)
    {
        double dist_x = turtle.x - pose_.x;
        double dist_y = turtle.y - pose_.y;
        return std::sqrt(dist_x * dist_x + dist_y * dist_y);
    }

    void callCatchTurtleService(std::string turtle_name)
    {
        while (!catch_turtle_client_->wait_for_service(1s))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for Service Server to be up...");
        }
        auto request = std::make_shared<my_robot_interfaces::srv::CatchTurtle::Request>() ;
        request->name = turtle_name ;

        catch_turtle_client_->async_send_request(request, 
            std::bind(&TurtleControllerNode::callbackCallCatchTurtleService, this, _1)) ;
        
    }

    void callbackCallCatchTurtleService(rclcpp::Client<my_robot_interfaces::srv::CatchTurtle>::SharedFuture future)
    {
        auto response = future.get() ;
        if (!response->success)
        {
            RCLCPP_ERROR(this->get_logger(), "Failed to remove turtle") ;
        }
    }

    void callbackTurtlepose(const turtlesim::msg::Pose::SharedPtr pose)
    {
        pose_ = *pose.get() ;
        turtlesim_up_ = true ;
    }

    void callbackAliveTurtles(const my_robot_interfaces::msg::TurtleArray::SharedPtr msg)
    {
        if (!msg->turtels.empty())
        {
            if (catch_closest_turtle_first_)
            {
                my_robot_interfaces::msg::Turtle closest_turtle = msg->turtels.at(0);
                double closest_turtle_distance = getDistanceFromCurrentPose(closest_turtle);

                for (int i = 1; i < (int)msg->turtels.size(); i++)
                {
                    double distance = getDistanceFromCurrentPose(msg->turtels.at(i));
                    if (distance < closest_turtle_distance)
                    {
                        closest_turtle = msg->turtels.at(i);
                        closest_turtle_distance = distance;
                    }
                }

                turtle_to_catch_ = closest_turtle;
            }
            else
            {
                turtle_to_catch_ = msg->turtels.at(0);
            }
        }
    }


    void PublishComdVel(double x, double theta)
    {
        auto msg = geometry_msgs::msg::Twist() ;
        msg.linear.x = x ;
        msg.angular.z = theta ;
        turtle_cmd_vel_->publish(msg) ;
    }

    void ControLoop()
    {
        if (!turtlesim_up_ || turtle_to_catch_.name == "")
        {
           return ;
        }
        double dist_x = turtle_to_catch_.x - pose_.x ;
        double dist_y = turtle_to_catch_.y - pose_.y ;
        double dist_ = sqrt(dist_x * dist_x + dist_y * dist_y) ;
         auto msg = geometry_msgs::msg::Twist() ;
        if (dist_ > 0.5)
        {
            msg.linear.x = 2*  dist_ ;
            double target_angle_ = atan2(dist_y, dist_x) ;
            double angle_ = target_angle_ - pose_.theta ;
            if (angle_ > M_PI)
                angle_ -= 2*M_PI ;
            else if (angle_ < -M_PI)
                 angle_ += 2*M_PI ;
            msg.angular.z = 6 * angle_ ;     
        }
        else
        {
            msg.linear.x = 0.0 ;
            msg.angular.z = 0.0 ;
            callCatchTurtleService(turtle_to_catch_.name) ;
            turtle_to_catch_.name = "" ;
        }
        turtle_cmd_vel_->publish(msg) ;
    }

    bool turtlesim_up_, catch_closest_turtle_first_ ;
    turtlesim::msg::Pose pose_ ;
    my_robot_interfaces::msg::Turtle turtle_to_catch_ ;

    rclcpp::Client<my_robot_interfaces::srv::CatchTurtle>::SharedPtr catch_turtle_client_ ;
    rclcpp::Subscription<my_robot_interfaces::msg::TurtleArray>::SharedPtr alive_turtles_subscrip_ ;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr  turtle_cmd_vel_ ;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle_pose_ ;
    rclcpp::TimerBase::SharedPtr control_loob_timer_ ;
}; 


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv) ;
    auto node = std::make_shared<TurtleControllerNode>() ;
    rclcpp::spin(node) ;
    rclcpp::shutdown() ; 
    return 0;
}

File: turtlesim_catch_py/turtlesim_catch_py/turtle_controller.py
import math
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
from functools import partial
from my_robot_interfaces.msg import Turtle
from my_robot_interfaces.msg import TurtleArray
from my_robot_interfaces.srv import CatchTurtle


class TurtleControllerNode(Node):
    def __init__(self):
        super().__init__("turtle_controller")
        self.turtle_to_catch : Turtle = None
        self.pose_: Pose = None
        # self.closest_turtle_
        self.alive_turtle_sub_ = self.create_subscription(
            TurtleArray, "/alive_turtles", self.callback_alive_turtles, 10)
        self.pose_subscriber_ = self.create_subscription(Pose, "/turtle1/pose", self.callback_pose, 10)
        self.cmd_publisher_ = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.control_time_ = self.create_timer(0.01, self.pub_cmd)
        self.catch_turtle_client_ = self.create_client(CatchTurtle, "/catch_turtle")

    def callback_call_CatchTurtle_service(self, future, name):
        response : CatchTurtle.Response = future.result()
        if not response.success :
            self.get_logger().error("Turtle " + name + " could not be removed")


    def call_catch_turtle_service(self,name):
        while not self.catch_turtle_client_.wait_for_service(1.0):
            self.get_logger().warn("wait for catch_turtle service...")
        
        request = CatchTurtle.Request()
        request.name = name

        future = self.catch_turtle_client_.call_async(request)
        future.add_done_callback(
            partial(self.callback_call_CatchTurtle_service, name = name))


    def callback_alive_turtles(self, live_turtle: TurtleArray):
        if len(live_turtle.turtels) > 0 :
            closest_turtle = None
            closest_turtle_distance = None
            for i in live_turtle.turtels :
                dist_x = i.x - self.pose_.x
                dist_y =i.y - self.pose_.y
                close_dis = math.sqrt(dist_x**2 + dist_y**2)
                if closest_turtle == None or close_dis < closest_turtle_distance :
                    closest_turtle_distance = close_dis
                    closest_turtle = i
                self.turtle_to_catch = closest_turtle

            # self.turtle_to_catch = live_turtle.turtels[0]


    def callback_pose(self, Pose: Pose):
        self.pose_ = Pose

    def pub_cmd(self):
        if self.pose_ == None or self.turtle_to_catch == None :
            return
        
        cmd = Twist()
        dist_x = self.turtle_to_catch.x -self.pose_.x 
        dist_y = self.turtle_to_catch.y - self.pose_.y 
        distance = math.sqrt(dist_x**2 + dist_y**2)
        
        if distance > .5 :
            cmd.linear.x = 2*distance

            theta = math.atan2(dist_y, dist_x)
            dif_theta = theta - self.pose_.theta 
            if dif_theta > math.pi :
                dif_theta -= 2*math.pi
            elif dif_theta < -math.pi :
                dif_theta += 2*math.pi
            cmd.angular.z = 6*dif_theta
        else :
            cmd.linear.x = 0.0
            cmd.angular.z = 0.0
            self.call_catch_turtle_service(self.turtle_to_catch.name)
            self.turtle_to_catch = None
        
        self.cmd_publisher_.publish(cmd)



def main(args=None):
    rclpy.init(args = args)
    node = TurtleControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__" :
    main()
